Zend/zend_compile.c

```diff
@@ -5711,6 +5717,74 @@ static zend_bool zend_is_valid_default_value(zend_type type, zval *value)
 	return 0;
 }
 
+static zend_attribute *zend_compile_attribute(zend_ast *ast, uint32_t offset) /* {{{ */
+{
+	ZEND_ASSERT(ast->kind == ZEND_AST_ATTRIBUTE);
+
+	zend_ast_list *list = ast->child[1] ? zend_ast_get_list(ast->child[1]) : NULL;
+	zend_attribute *attr = emalloc(ZEND_ATTRIBUTE_SIZE(list ? list->children : 0));
+
+	attr->name = zend_resolve_class_name_ast(ast->child[0]);
+	attr->lcname = zend_string_tolower(attr->name);
+	attr->offset = offset;
+	attr->argc = list ? list->children : 0;
+
+	if (ast->child[1]) {
+		ZEND_ASSERT(ast->child[1]->kind == ZEND_AST_ARG_LIST);
+
+		uint32_t i;
+
+		for (i = 0; i < list->children; i++) {
+			zend_const_expr_to_zval(&attr->argv[i], list->child[i]);
+		}
+	}
+
+	return attr;
+}
+/* }}} */
+
+static void attribute_ptr_dtor(zval *v) /* {{{ */
+{
+	zend_attribute_free((zend_attribute *) Z_PTR_P(v));
+}
+/* }}} */
+
+static zend_always_inline HashTable *create_attribute_array(uint32_t size) /* {{{ */
```

Comment:
I think you can safely drop the `size` argument here and make the caller code a bit nicer. The minimum hashtable size is 8, and I don't think that having more than 8 attributes on one syntactic element is common enough that we need to bother optimizing for it.